### ISSUE TITLE
front-end: Export `dynamic` from `dart:core`.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/sdk#32228]
The `dynamic` type should be should be exported by the `dart:core` library under the `dynamic` identifier.

As a built-in identifier, it's possible to declare members named `dynamic` (even thought you probably shouldn't), and those will shadow the `dynamic` in `dart:core`.

See #29125.

**Comment by sigmundch** [comment]
This will be handled entirely in the common front end. Reassigning to the fe team.

**Comment by mkustermann** [comment]
Is there any updates on this? (@jakobr-google )

**Comment by leafpetersen** [comment]
There is a bit of an unfortunate state right now in that as far as I can tell there's no way to write code that prefixes `dart:core` and references `dynamic` and works on both the CFE and the analyzer.  

Even if the main issue (and it's accompanying analyzer issue) is lower priority, I do feel like we should arrange things so that we don't have this catch-22.  cc @kmillikin @bwilkerson 

```
import "dart:core" as core;

void main() {
  core.dynamic a; // Analyzer error
  dynamic b;  // CFE error
}
```

**Comment by dgrove** [comment]
What is the concrete work that needs to happen here, @leafpetersen @bwilkerson ?

**Comment by lrhn** [comment]
The analyzer needs to consider `dynamic` a normal (built-in) identifier in all contexts, and ensure that dart:core exports this identifier as a reference to the dynamic type.

Example code:
```dart
import "dart:core" hide dynamic;                                                    
import "dart:core" as core show dynamic;

core.dynamic get x => 42;
int dynamic = 10;
main() {
  print(dynamic);  // prints 10
  print(x is core.dynamic);  // prints true.
}
```
This code runs correctly on the current VM and dart2js, with and without `--preview-dart-2`.
It gives the following errors in the analyzer:
>   error • Undefined class 'core.dynamic' at dyn.dart:4:1 • undefined_class          
>  error • The name 'core.dynamic' isn't defined, so it can't be used in an 'is' expression at dyn.dart:10:14 • type_test_with_undefined_name                             
>  hint • The library 'dart:core' doesn't export a member with the hidden name 'dynamic' at dyn.dart:1:25 • undefined_hidden_name                                         
>  hint • The library 'dart:core' doesn't export a member with the shown name 'dynamic' at dyn.dart:2:33 • undefined_shown_name                                           
> 2 errors and 2 hints found.

**Comment by mkustermann** [comment]
This problem is in fact **causing ongoing issues** when rolling `package:googleapis` into google3 because our API generator generates prefix-imported `import "dart:core" as core;` and uses `core.dynamic` (which is perfectly valid).

We have to **manually edit** the code we check into google3 to make the presubmit checks pass.

**Comment by dgrove** [comment]
This should be closed, correct (since #32227 is where there is remaining work to do)?

**Comment by JekCharlsonYu** [comment]
closing as per discussion with Kevin/Dan
